processing/src/main/java/org/apache/druid/segment/filter/ExpressionFilter.java

```diff
@@ -55,6 +69,105 @@ public ExpressionFilter(final Supplier<Expr> expr, final FilterTuning filterTuni
     this.filterTuning = filterTuning;
   }
 
+  @Override
+  public boolean canVectorizeMatcher(ColumnInspector inspector)
+  {
+    return expr.get().canVectorize(inspector);
+  }
+
+  @Override
+  public VectorValueMatcher makeVectorMatcher(VectorColumnSelectorFactory factory)
+  {
+    final Expr theExpr = expr.get();
+
+    DruidPredicateFactory predicateFactory = new DruidPredicateFactory()
+    {
+      @Override
+      public Predicate<String> makeStringPredicate()
+      {
+        return Evals::asBoolean;
+      }
+
+      @Override
+      public DruidLongPredicate makeLongPredicate()
+      {
+        return Evals::asBoolean;
+      }
+
+      @Override
+      public DruidFloatPredicate makeFloatPredicate()
+      {
+        return Evals::asBoolean;
+      }
+
+      @Override
+      public DruidDoublePredicate makeDoublePredicate()
+      {
+        return Evals::asBoolean;
+      }
+
+      // The hashcode and equals are to make SubclassesMustOverrideEqualsAndHashCodeTest stop complaining..
+      // DruidPredicateFactory doesn't really need equals or hashcode, in fact only the 'toString' method is called
```

Comment:
Interesting. It sounds like a bug that DruidPredicateFactory is annotated with `@SubclassesMustOverrideEqualsAndHashCodeTest` but yet its toString method is used for equality checks. Is that right? If so, perhaps a good fix would be:

- Keep the annotation on DimensionPredicateFilter
- Switch the equality checks in DimensionPredicateFilter to use DruidPredicateFactory `equals` rather than `toString`

If we do that, IMO it'd be good to do it separately from this patch and also separately from #8256.

> DruidPredicateFactory doesn't really need equals or hashcode

How else would DimensionPredicateFilter's `equals` be implemented?

Reply:
yeah, I was planning on addressing this in a separate PR, the comments are partially to help remind us to do it in case I didn't get to it immediately

Reply:
👍  although I'm still wondering: how else would DimensionPredicateFilter's equals be implemented, if not via equals and hashCode on DruidPredicateFactory?

(Assuming we stop it from using toString, which IMO we should)

Reply:
oops, I forgot to rework this comment, it is assuming that it keeps using toString. If we changed that to not use the string for its equality then it would be the toString that isn't necessary, will try to rework

Reply:
changed comment to suggest `DimensionPredicateFilter` should use equals of predicate factory instead of string form